src/mscorlib/src/System/String.Manipulation.cs

```diff
@@ -585,17 +642,72 @@ public static string Join(string separator, params object[] values)
         }
 
         [ComVisible(false)]
-        public static String Join<T>(String separator, IEnumerable<T> values)
+        public unsafe static string Join<T>(string separator, IEnumerable<T> values)
         {
+            separator = separator ?? string.Empty;
+            fixed (char* pSeparator = &separator.m_firstChar)
+            {
+                // Defer argument validataion to the internal function
+                return JoinCore(pSeparator, separator.Length, values);
+            }
+        }
+
+        [ComVisible(false)]
+        public static String Join(String separator, IEnumerable<String> values) {
             if (values == null)
                 throw new ArgumentNullException(nameof(values));
             Contract.Ensures(Contract.Result<String>() != null);
             Contract.EndContractBlock();
 
+            using(IEnumerator<String> en = values.GetEnumerator()) {
+                if (!en.MoveNext())
+                    return String.Empty;
+
+                String firstValue = en.Current;
+
+                if (!en.MoveNext()) {
+                    // Only one value available
+                    return firstValue ?? String.Empty;
+                }
+
+                // Null separator and values are handled by the StringBuilder
+                StringBuilder result = StringBuilderCache.Acquire();
+                result.Append(firstValue);
+
+                do {
+                    result.Append(separator);
+                    result.Append(en.Current);
+                } while (en.MoveNext());
+                return StringBuilderCache.GetStringAndRelease(result);
+            }
+        }
+
+        // Joins an array of strings together as one string with a separator between each original string.
+        //
+        [System.Security.SecuritySafeCritical]  // auto-generated
```

Comment:
@jamesqo that would work. I was just thinking that since you are adding this method you can clean-up the comment. I am fine with it as-is :).